### PR TITLE
Feat: [HL-25] add validator with the ability to validate email & password✌️

### PR DESCRIPTION
**Pull request by claramounir**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature

## Description
add validator with the ability to validate email & password

## Added tests?

- [x] 🙅 no, because they aren't needed

## Added to documentation?

- [x] 🙅 no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?
https://media.tenor.com/_xzAZjgTH2wAAAAd/joey-richter-starkid.gif

### CODE REVIEW

File: Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+//
+
+import Foundation
+public enum HealthyValidationRule {
+
+    case matching(value: String, regex: String?)
+    case numericOnly(value: String)
+    case decimal(value: String)
+    case charCount(value: String, minCount: Int?, maxCount: Int?)
```
Comment (abdelhamid-f-nasser) [inline Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift:14]: I think can split this to 3 cases, `minRequiredCharCount`, `maxRequiredCharCount`, `rangeRequiredCharCount` to make it more flexible for the future changes.

File: Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    var validationError: String? {
+        if !isValid {
+            switch self {
+            case .matching(value: _, regex: _):
+                return "incorrect Format"
+            case .numericOnly(value: _):
+                return "numeric Error"
+            case .decimal(value: _):
+                return "not a Number"
```
Comment (abdelhamid-f-nasser) [inline Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift]: We can change this to "Input is not a valid number" or "Please enter a valid number"

File: Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            case .firstChar(value: _, chars: let chars):
+                let stringArray = chars.map({ return String($0) })
+                let prefixes = stringArray.joined(separator: ", ")
+                return  "incorrect Prefix \(prefixes)"
+
+            case .withinLimits(value: let value, min: let min, max: let max):
+                guard let doubleValue = Double(value) else { return nil }
+
+                if doubleValue > max {
+                    return  "incorrect Amount Max"
```
Comment (abdelhamid-f-nasser) [inline Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift]: We can change this to "value exceeds maximum allowed value" for better error readability or even show the max allowed value in the error.

File: Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                } else if doubleValue < min {
+                    return  "incorrect Amount Min"
```
Comment (abdelhamid-f-nasser) [inline Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift]: We can change this to "value fails to reach the minimum allowed value" for better error readability or even show the min allowed value in the error.

File: Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidator .swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
+//
+//  HealthyValidator .swift
+//  Healthy
+//
+//
+
+import Foundation
+public protocol HealthyValidator {
```
Comment (abdelhamid-f-nasser) [inline Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidator .swift:8]: Great job making this extensible !!!  [LGTM](https://media.giphy.com/media/NytMLKyiaIh6VH9SPm/giphy.gif)

**Review by abdelhamid-f-nasser — COMMENTED** [review]
Great job making the `HealthyValidator ` extensible !!!  

LGTM

File: Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift (hunk 1 above)
Comment (claramounir) [reply Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift:14]: in this case, we have minimum  and maximum counts as an optional value so you can send them with a nil value to satisfy the 3 cases you mentioned (ex: if you want only minimum you should send the maximum value with nil )

**Review by abdelhamid-f-nasser — APPROVED** [review]
Well done, you did a great job making the error messages more readable !  👍🏼

**Review by mhmdatallaa — APPROVED** [review]
Great work ya Clara 💫

**Pull request merged**